library/core/src/hint.rs

```diff
@@ -152,23 +152,79 @@ pub fn spin_loop() {
 /// backend used. Programs cannot rely on `black_box` for *correctness* in any way.
 ///
 /// [`std::convert::identity`]: crate::convert::identity
-#[cfg_attr(not(miri), inline)]
-#[cfg_attr(miri, inline(never))]
+#[inline]
 #[unstable(feature = "bench_black_box", issue = "64102")]
-#[cfg_attr(miri, allow(unused_mut))]
+#[allow(unused_mut)]
 pub fn black_box<T>(mut dummy: T) -> T {
     // We need to "use" the argument in some way LLVM can't introspect, and on
     // targets that support it we can typically leverage inline assembly to do
     // this. LLVM's interpretation of inline assembly is that it's, well, a black
     // box. This isn't the greatest implementation since it probably deoptimizes
     // more than we want, but it's so far good enough.
-
-    #[cfg(not(miri))] // This is just a hint, so it is fine to skip in Miri.
+    #[cfg(all(
+        not(miri),
+        any(
+            target_arch = "aarch64",
+            target_arch = "arm",
+            target_arch = "x86",
+            target_arch = "x86_64",
+            target_arch = "riscv32",
+            target_arch = "riscv64",
+            target_arch = "nvptx64",
+            target_arch = "powerpc",
+            target_arch = "powerpc64",
+            target_arch = "hexagon",
+            target_arch = "mips",
+            target_arch = "mips64",
+            target_arch = "spirv",
+            target_arch = "wasm32",
+            target_arch = "bpf",
+        )
+    ))]
+    #[allow(asm_sub_register)]
     // SAFETY: the inline assembly is a no-op.
     unsafe {
-        // FIXME: Cannot use `asm!` because it doesn't support MIPS and other architectures.
-        llvm_asm!("" : : "r"(&mut dummy) : "memory" : "volatile");
+        // wasm32 uses local instead of reg
+        #[cfg(not(target_arch = "wasm32"))]
+        asm!("/* {} */", in(reg) &mut dummy, options(nostack, preserves_flags));
+        #[cfg(target_arch = "wasm32")]
+        asm!("/* {} */", in(local) &mut dummy, options(nostack, preserves_flags));
+        dummy
     }
 
-    dummy
+    // Use volatile operations on targets that don't support inline assembly.
+    // This effectively simulates:
+    // 1. passing the value to some external process
+    // 2. allowing that process to access any memory that the value may point to
+    // 3. reading a potentially modified value back
+    #[cfg(any(
```

Comment:
My concern with this implementation is that once support for assembly is added for any of these architectures, it is trivial to miss updating this location.

Any way we can avoid this? We _could_ make up an intrinsic for this, perhaps?

Reply:
Instead of using the fallback for all *other* architectures, could we have an explicit list of architectures that *do* use the fallback?

Reply:
> Instead of using the fallback for all other architectures, could we have an explicit list of architectures that do use the fallback?

This will likely cause problems when people try to use a new architecture with a custom json target.

> Any way we can avoid this? We could make up an intrinsic for this, perhaps?

We could have a `target_has_asm` cfg flag like we do for atomics. But it seems a bit excessive to have this just for `black_box`.

Reply:
Having thought about this more, an intrinsic seems significantly more straightforward to me. But I guess I'm okay with this implementation too.

If we continue using cfgs here, consider `cfg_if!` maybe? It'll reduce the clutter here significantly I feel.